### PR TITLE
[BOOST-4787] feat: extend erc20variableincentive to include on chain criteria

### PR DESCRIPTION
**Pull request by Quazia**
🚨 Please review the [guidelines for contributing](https://github.com/boostxyz/boost-protocol/blob/main/.github/CONTRIBUTING.md) to this repository.

### Description
Adds the contract and ts helper files to facilitate variable incentive criteria on-chain

💔 Thank you!

### CODE REVIEW

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Are you sure you want to be submitting a change without including a changeset? If you're just changing docs or tests, you probably don't need to. See [the publishing section of the README](https://github.com/boostxyz/boost-protocol?tab=readme-ov-file#changesets--publishing) for more info.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 5c5ebd4308d6c48141e74893eb8133e758590537
</p>

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Are you sure you want to be submitting a change without including a changeset? If you're just changing docs or tests, you probably don't need to. See [the publishing section of the README](https://github.com/boostxyz/boost-protocol?tab=readme-ov-file#changesets--publishing) for more info.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 78210519df264ac4fd998cd1e396c031b90ab8af
</p>

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Are you sure you want to be submitting a change without including a changeset? If you're just changing docs or tests, you probably don't need to. See [the publishing section of the README](https://github.com/boostxyz/boost-protocol?tab=readme-ov-file#changesets--publishing) for more info.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 01dcc4af8f523e9460886d39479e27d2bda860c6
</p>

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Are you sure you want to be submitting a change without including a changeset? If you're just changing docs or tests, you probably don't need to. See [the publishing section of the README](https://github.com/boostxyz/boost-protocol?tab=readme-ov-file#changesets--publishing) for more info.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 29db578fbe5726673c166d2f380519e5056a2302
</p>

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Are you sure you want to be submitting a change without including a changeset? If you're just changing docs or tests, you probably don't need to. See [the publishing section of the README](https://github.com/boostxyz/boost-protocol?tab=readme-ov-file#changesets--publishing) for more info.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 101e8323a3ce93731f7b39520914a3e07f1364ff
</p>

File: packages/sdk/src/Incentives/ERC20VariableCriteriaIncentive.ts [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+}
+
+class NoMatchingLogsError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = 'NoMatchingLogsError';
+  }
+}
+
+class InvalidCriteriaTypeError extends Error {
```
[the remark this answers is not in the document]
Comment (Quazia) [reply packages/sdk/src/Incentives/ERC20VariableCriteriaIncentive.ts:99]: Nah I can handle this I'll put them in the right folder and clean them up I just wanted a placeholder so I didn't use non custom errors

File: packages/sdk/src/Incentives/ERC20VariableCriteriaIncentive.ts [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+  > {}
+
+export interface GetIncentiveScalarParams {
+  hash: Hex;
+}
+
+/**
+ * Extended ERC20 Variable Criteria Incentive class that fetches incentive criteria and scalar
+ */
+export class ERC20VariableCriteriaIncentive extends DeployableTarget<
```
[the remark this answers is not in the document]
Comment (Quazia) [reply packages/sdk/src/Incentives/ERC20VariableCriteriaIncentive.ts:126]: Cool I'll look what you did there - yeah I want to be able to pass through the types for the deployable

File: packages/sdk/src/Incentives/ERC20VariableCriteriaIncentive.ts [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+      );
+    }
+  }
+
+  /**
+   * Fetches the incentive scalar from a transaction hash
+   *
+   * @param {GetIncentiveScalarParams} params
+   * @returns {Promise<bigint>}
+   * @throws {InvalidCriteriaTypeError | NoMatchingLogsError | DecodedArgsError}
```
[the remark this answers is not in the document]
Comment (Quazia) [reply packages/sdk/src/Incentives/ERC20VariableCriteriaIncentive.ts:176]: I think I prefer the union but it's trivial to change and I'm loosely opinionated.

File: packages/evm/contracts/incentives/ERC20VariableCriteriaIncentive.sol [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        InitPayloadExtended memory init_ = abi.decode(data_, (InitPayloadExtended));
+
+        address asset_ = init_.asset;
+        uint256 reward_ = init_.reward;
+        uint256 limit_ = init_.limit;
+        IncentiveCriteria memory criteria_ = init_.criteria;
+
+        if (limit_ == 0) revert BoostError.InvalidInitialization();
+
+        uint256 available = asset_.balanceOf(address(this));
```
[the remark this answers is not in the document]
Comment (Quazia) [reply packages/evm/contracts/incentives/ERC20VariableCriteriaIncentive.sol:52]: [It's this function where the magic happens](https://github.com/boostxyz/boost-protocol/blob/main/packages/evm/contracts/BoostCore.sol#L278)

[We preflight the value disbursal here](https://github.com/boostxyz/boost-protocol/blob/main/packages/evm/contracts/BoostCore.sol#L294-L298)

[Then initialize after that](https://github.com/boostxyz/boost-protocol/blob/main/packages/evm/contracts/BoostCore.sol#L301)

You can see from the tests too we're initializing it correctly.

**Comment by sammccord** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>

  Are you sure you want to be submitting a change without including a changeset? If you're just changing docs or tests, you probably don't need to. See [the publishing section of the README](https://github.com/boostxyz/boost-protocol?tab=readme-ov-file#changesets--publishing) for more info.
  </td>
    </tr>
  </tbody>
</table>



<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against acf528985cab3b9085d3d2c3fc162555c9b63e7c
</p>

**Review by sammccord — APPROVED** [review]

**Pull request merged**